### PR TITLE
Add pre-commit

### PR DESCRIPTION
**Pull request by brenns10**
In my past Python projects I've relied on pre-commit hooks to catch lint errors before I even push my branch to CI. I found this to be super useful (and have already re-pushed quite a few drgn commits after realizing I forgot to run black / isort / mypy).

This PR adds support for the pre-commit tool, including the equivalent checks which are run during test/lint stages. Developers can install it locally with `pre-commit install --install-hooks`, where it will check your changed files prior to each commit. I ran the pre-commit hooks across all files and caught one black reformatting (which may have been my fault anyway...), which I included here. Finally, I updated the CI steps to directly call pre-commit for linting.

This means that all linter versions and arguments can be stored in a single place (`.pre-commit-config.yaml`), and since the versions are pinned, the lint errors should be reproducible locally and on CI.

Consider this closer to an RFC. It's ready to merge but I don't know if it's a direction you're interested in going.

### CODE REVIEW

**Comment by brenns10** [comment]
pre-commit supports lots of other checks too:

- flake8 which can automatically check things like unused variables
- end-of-file-fixer to ensure a newline always exists at EOF
- a fixer for trailing whitespace

I've even been able to get pre-commit to enforce C coding style guidelines similar to the Linux kernel's, however it would depend on clang-format which is not ideal. (here's my [pre-commit config](https://git.sr.ht/~brenns10/sc-template/tree/master/item/.pre-commit-config.yaml) and [clang-format config](https://git.sr.ht/~brenns10/sc-template/tree/master/item/.clang-format))

And [even more](https://pre-commit.com/hooks.html) options exist on pre-commit's documentation.

I'm not including anything like that because I want this to essentially be a drop-in replacement for the current linter setup, to discuss the merits of pre-commit. We can discuss additional hooks later if we decide to go this route.

**Review by osandov — COMMENTED** [review]
This is interesting. I'm not 100% sure about it, so here are my thoughts. I'd love to hear your opinion on these points:

Things I like:

* The centralized configuration for linters.
* Ease of adding new checks.

Things I don't like:

* The hooks modify files rather than running in a "check only" mode. This doesn't matter for the CI, but for local use that seems like it'd be really annoying. For example, you'd have to reload files you have open after committing. I could also see this being confusing with interactive patch commits (git commit -p), which I do often.

Things I'm not sure about:

* Version pinning of the linters. On the one hand, the reproducibility is good, and I've definitely had issues with new versions creating new errors. On the other hand, keeping up to date with new releases and manually updating the version would be a maintenance burden.

**Comment by brenns10** [comment]
Here are my thoughts:

> Things I like:
>
> - The centralized configuration for linters.
> - Ease of adding new checks.

Especially the ease of adding new checks is great! It's nice to be able to browse the pre-commit.com list of hooks as a bit of a "shopping list" for your project. But at the same time each new hook can add overhead to CI and friction when ramping up new contributors. So there's a balance. But it's nice to have just one file to manage all of the linting.

> Things I don't like:
>
> - The hooks modify files rather than running in a "check only" mode. This doesn't matter for the CI, but for local use that seems like it'd be really annoying. For example, you'd have to reload files you have open after committing. I could also see this being confusing with interactive patch commits (git commit -p), which I do often.

I totally can see that concern. I've used pre-commit for several years, and I would be lying if I said that there was never a confusing or frustrating moment. The way it works is that it will transparently stash any unstaged modifications, and only run the hooks on the changed files. At the end, regardless of the success of the hooks, it will restore the stashed modifications. If the stashed modifications conflict with the changes by a hook, then it simply drops the changes from the hook, in favor of your stashed changes. Here's a trivial example where I staged `blah.py` with out-of-order imports, then further modified it to contain another import:

```
$ git diff --cached
diff --git a/blah.py b/blah.py
new file mode 100644
index 0000000..5bc6568
--- /dev/null
+++ b/blah.py
@@ -0,0 +1,7 @@
+from collections import Counter
+from abc import ABC
+
+
+class Base(ABC):
+    def __init__(self, *args):
+        self.counter = Counter(args)

$ git diff
diff --git a/blah.py b/blah.py
index 5bc6568..349e05c 100644
--- a/blah.py
+++ b/blah.py
@@ -1,7 +1,9 @@
 from collections import Counter
 from abc import ABC
+import socket
 
 
 class Base(ABC):
     def __init__(self, *args):
         self.counter = Counter(args)
+        self.socket = socket.socket()
$ git commit
[WARNING] Unstaged files detected.
[INFO] Stashing unstaged files to /home/stepbren/.cache/pre-commit/patch1641937912-240847.
isort (python)...........................................................Failed
- hook id: isort
- files were modified by this hook

Fixing /home/stepbren/repos/drgn/blah.py

black....................................................................Passed
mypy.................................................(no files to check)Skipped
[WARNING] Stashed changes conflicted with hook auto-fixes... Rolling back fixes...
[INFO] Restored changes from /home/stepbren/.cache/pre-commit/patch1641937912-240847.
```

You can see that your changes are always respected, and hopefully it isn't too confusing. Given that `git commit -p` is just shorthand for running `git add -p && git commit`, whatever you don't select for committing would be stashed by pre-commit, and would be restored after the commit. If the hooks' changes conflict with your unstaged changes, your changes win out.

**However,** there's a way around all this. pre-commit leaves the decision up to the hook to determine whether it changes files. Most common linters have a `--diff` or `--check` option so that they only report errors, or suggest diffs without actually modifying files. In the pre-commit config provided here, isort and black default to modifying the code, whereas mypy only reports errors. I looked at the docs, and we just need to provide `--check --diff` for isort, and `--diff` for black, and we'll get the no-modify behavior that you prefer. Here's how pre-commit behaves on the same example, if we add those arguments:

```
$ git commit
[WARNING] Unstaged files detected.
[INFO] Stashing unstaged files to /home/stepbren/.cache/pre-commit/patch1641939831-243288.
isort (python)...........................................................Failed
- hook id: isort
- exit code: 1

ERROR: /home/stepbren/repos/drgn/blah.py Imports are incorrectly sorted and/or formatted.
--- /home/stepbren/repos/drgn/blah.py:before    2022-01-11 14:23:51.668761
+++ /home/stepbren/repos/drgn/blah.py:after     2022-01-11 14:23:51.756558
@@ -1,5 +1,5 @@
+from abc import ABC
 from collections import Counter
-from abc import ABC
 
 
 class Base(ABC):

black....................................................................Passed
mypy.................................................(no files to check)Skipped
[INFO] Restored changes from /home/stepbren/.cache/pre-commit/patch1641939831-243288.
```

The only weird thing here is that the hook output doesn't reflect our unstaged changes, but that's necessary to properly lint "just this commit".

> Things I'm not sure about:
>
> - Version pinning of the linters. On the one hand, the reproducibility is good, and I've definitely had issues with new versions creating new errors. On the other hand, keeping up to date with new releases and manually updating the version would be a maintenance burden.

Yes, it's a bit of additional maintenance, but thankfully pre-commit has `pre-commit autoupdate`. This will go ahead and find the latest tag for each hook, updating to the latest. You still need to remember to run it once in a while, but it's not like you need to pop open 3 github pages, find the most recent release, and manually update the `.pre-commit-config.yaml` file.

For the projects I've used pre-commit with, I find that the benefit of reproducible linting outweighs the burden of running `pre-commit autoupdate`. Especially because it's good practice to run `pre-commit run --all-files` once you've updated hooks: how else can you detect if the linters have new rules that your code "violates"? :laughing: 

---

In summary I think your concerns are totally valid but hopefully manageable. If you agree, I'll fix this up with the latest hook versions, and handle your other CR comments. If not I can continue to use pre-commit locally! :grin:

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -23,11 +23,11 @@ jobs:
         run: |
           sudo apt-get update
           sudo apt-get install busybox-static libelf-dev libdw-dev qemu-kvm zstd ${{ matrix.cc == 'clang' && 'libomp-$(clang --version | sed -rn "s/.*clang version ([0-9]+).*/\\1/p")-dev' || '' }}
-          pip install mypy pyroute2
+          pip install mypy pyroute2 pre-commit
```
[the remark this answers is not in the document]
Comment (brenns10) [reply .github/workflows/ci.yml:26]: pre-commit sets up its own version, so we can remove mypy from this list.

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -38,8 +38,6 @@ jobs:
...
         uses: actions/setup-python@v2
       - name: Install dependencies
-        run: pip install black isort
-      - name: Check with black
-        run: black --check --diff .
-      - name: Check with isort
-        run: isort --check --diff .
+        run: pip install pre-commit
+      - name: Run pre-commit hooks
+        run: pre-commit run --all-files --show-diff-on-failure
```
[the remark this answers is not in the document]
Comment (brenns10) [reply .github/workflows/ci.yml:43]: I didn't know about the SKIP environment variable! Yeah, this makes sense. That way if there are mypy failures AND other linter failures, you won't neglect to look at the output here because you see the same mypy warnings from the other steps.

File: .pre-commit-config.yaml [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+        name: isort (python)
+-   repo: https://github.com/psf/black
+    rev: 21.12b0
+    hooks:
+    -   id: black
+-   repo: https://github.com/pre-commit/mirrors-mypy
+    rev: v0.921
+    hooks:
+    -   id: mypy
+        args: [--no-strict-optional, --ignore-missing-imports]
```
[the remark this answers is not in the document]
Comment (brenns10) [reply .pre-commit-config.yaml:15]: I do not know where I got those from :rofl:

**Comment by osandov** [comment]
Thanks for the detailed response! Given your explanation of how `pre-commit` handles conflicts, and the fact that I could manually `pre-commit run` when needed rather than using it as an actual pre-commit hook if I still don't like it, I think I'm fine with the modification behavior after all. `pre-commit autoupdate` also assuages my concern about version pinning. So let's go ahead and try it :)

**Comment by brenns10** [comment]
Great! I went ahead with the changes.

**Comment by osandov** [comment]
Merged, thank you!
